app/code/Magento/InventoryShipping/Plugin/Sales/Shipment/AssignSourceCodeToShipmentPlugin.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryShipping\Plugin\Sales\Shipment;
+
+use Magento\Framework\App\RequestInterface;
+use Magento\Sales\Api\Data\ShipmentInterface;
+use Magento\Sales\Model\Order;
+use Magento\Sales\Model\Order\ShipmentFactory;
+use Magento\Sales\Api\Data\ShipmentExtensionFactory;
+use Magento\InventorySales\Model\StockByWebsiteIdResolver;
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\InventoryApi\Api\GetStockSourceLinksInterface;
+use Magento\InventoryApi\Api\Data\StockSourceLinkInterface;
+
+class AssignSourceCodeToShipmentPlugin
+{
+    /**
+     * @var RequestInterface
+     */
+    private $request;
+
+    /**
+     * @var ShipmentExtensionFactory
+     */
+    private $shipmentExtensionFactory;
+
+    /**
+     * @var StockByWebsiteIdResolver
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var GetStockSourceLinksInterface
+     */
+    private $getStockSourceLinks;
+
+    /**
+     * AssignSourceCodeToShipmentPlugin constructor.
+     * @param RequestInterface $request
+     * @param ShipmentExtensionFactory $shipmentExtensionFactory
+     * @param StockByWebsiteIdResolver $stockByWebsiteIdResolver
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param GetStockSourceLinksInterface $getStockSourceLinks
+     */
+    public function __construct(
+        RequestInterface $request,
+        ShipmentExtensionFactory $shipmentExtensionFactory,
+        StockByWebsiteIdResolver $stockByWebsiteIdResolver,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        GetStockSourceLinksInterface $getStockSourceLinks
+    ) {
+        $this->request = $request;
+        $this->shipmentExtensionFactory = $shipmentExtensionFactory;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->getStockSourceLinks = $getStockSourceLinks;
+    }
+
+    /**
+     * @param ShipmentFactory $subject
+     * @param ShipmentInterface $shipment
+     * @param Order $order
+     * @return ShipmentInterface
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterCreate(ShipmentFactory $subject, ShipmentInterface $shipment, Order $order)
+    {
+        $sourceCode = $this->request->getParam('source_code');
+        if (empty($sourceCode)) {
```

Comment:
you need somehow to describe that in code that in this case, you are dealing with Single Source mode.
For example, calling `\Magento\InventoryCatalog\Model\IsSingleSourceModeInterface` service.

Because currently, this code looks awkward
```php
            $websiteId = $order->getStore()->getWebsiteId();
            $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
            $sources = $this->getAssignedSourcesForStock($stockId);
            if (!empty($sources) && count($sources) == 1) {
                $sourceCode = $sources[0];
            }
```

because if we are in Multi-Source mode and `$sourceCode` is empty - this is an exceptional case and we need to throw exception